nibabel/bv.py

```diff
@@ -0,0 +1,738 @@
+    The parameter ``proto`` descibed below is notated accoring to
+    https://docs.python.org/3/reference/introduction.html#notation
+    (except using double parens indicating a group, single parens indicating
+    a tuple).
+
+    elemement_proto ::= (name, pack-format, default)  |
+                        name, (n_fields_name, <element_proto>)
+    proto ::= ((element_proto))*
+    """
+    hdr_dict = OrderedDict()
+    for k in proto:
+        hdr_dict[k[0]] = {}
+        if type(k[1]) is tuple:
+            hdr_dict[k[0]]['dt'] = 'multi'
+            hdr_dict[k[0]]['nField'] = k[1][0]
+            hdr_dict[k[0]]['default'] = hdr_dict[k[0]]['value'] = \
+                [_make_hdr_dict((k[1][1],))]
+        else:
+            hdr_dict[k[0]]['dt'] = k[1]
+            hdr_dict[k[0]]['default'] = hdr_dict[k[0]]['value'] = k[2]
+    return hdr_dict
+
+def readCString(f, nStrings=1, bufsize=1000, startPos=None, strip=True,
+                rewind=False):
+    """Read a zero-terminated string from a file object.
+
+    Read and return a zero-terminated string from a file object.
+
+    Parameters
+    ----------
+    f : fileobj
+       File object to use
+    nStrings: int, optional
+       Number of strings to search (and return). Default is 1.
+    bufsize: int, optional
+       Define the buffer size that should be searched for the string.
+       Default is 1000 bytes.
+    startPos: int, optional
+       Define the start file position from which to search. If None then start
+       where the file object currently points to. Default is None.
+    strip : bool, optional
+       Whether to strip the trailing zero from the returned string.
+       Default is True.
+    rewind: bool, optional
+       Whether the fileobj f should be returned to the initial position after
+       reading. Default is False.
+
+    Returns
+    -------
+    str_list : generator of string(s)
+    """
+    currentPos = f.tell()
+    if strip:
+        suffix = ''
+    else:
+        suffix = '\x00'
+    if startPos is not None:
+        f.seek(startPos)
+    data = f.read(bufsize)
+    lines = data.split('\x00')
+    str_list = []
+    if rewind:
+        f.seek(currentPos)
+    else:
+        offset = 0
+        for s in range(nStrings):
+            offset += len(lines[s])+1
+        f.seek(currentPos+offset)
+    for s in range(nStrings):
+        str_list.append(lines[s] + suffix)
+    return str_list
+
+def parse_BV_header(hdr_dict, fileobj):
+    """Parse the header of a BV file format.
+
+    This function can be (and is) called recursively to iterate through nested
+    fields (e.g. the ``prts`` field of the VTC header).
+
+    Parameters
+    ----------
+    hdrDict: OrderedDict
+       Pre-populated hdrDict that contains the fields to expect for the
+       respective BV file format.
+    fileobj : fileobj
+       File object to use. Make sure that the current position is at the
+       beginning of the header (e.g. at 0).
+    """
+    for key in hdr_dict:
```

Comment:
Getting rid of default would not allow me to create valid empty ('new') BV files. For the same reason I though it is nice to separate ```_make_hdr_dict``` and ```parse_BV_header``` because one would only generate a default header from a given prototype and the other would be used for actual parsing. I can still of course combine it.
Getting rid of dt would break parsing and/or writing as some header fields have different datatypes for numbers (_e.g._ ```RefSpace``` in ```VTC``` files). Or how would you solve this without ```dt```?